capsules/src/hmac.rs

```diff
@@ -0,0 +1,492 @@
+                            // Determine the size of the static buffer we have
+                            let static_buffer = buf.len();
+
+                            // If we have more data then the static buffer we set how much data we are going to copy
+                            if d.as_ref().len() > static_buffer {
+                                self.data_copied.set(static_buffer);
+                            }
+
+                            // Copy the data into the static buffer
+                            if d.as_ref().len() < static_buffer {
+                                // If the static buffer is larger then the data, just copy the full slice
+                                buf.copy_from_slice(d.as_ref());
+                            } else {
+                                // Copy the entire buffers worth of data
+                                buf.copy_from_slice(d.as_ref().get(0..static_buffer).unwrap());
+                            }
+                        });
+
+                        // Add the data from the static buffer to the HMAC
+                        if let Err(e) = self
+                            .hmac
+                            .add_data(LeasableBuffer::new(self.data_buffer.take().unwrap()))
+                        {
+                            self.data_buffer.replace(e.1);
+                            return e.0;
+                        }
+                    }
+                    None => {
+                        return ReturnCode::ERESERVE;
+                    }
+                };
+
+                ReturnCode::SUCCESS
+            })
+            .unwrap_or_else(|err| err.into())
+    }
+}
+
+impl<H: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> digest::Client<'a, T>
+    for HmacDriver<'a, H, T>
+{
+    fn add_data_done(&'a self, _result: Result<(), ReturnCode>, data: &'static mut [u8]) {
+        self.appid.map(move |id| {
+            self.apps
+                .enter(*id, move |app, _| {
+                    let mut data_len = 0;
+                    let mut static_buffer = 0;
+
+                    self.data_buffer.replace(data);
+
+                    self.data_buffer.map(|buf| {
+                        match app.data.as_ref() {
+                            Some(d) => {
+                                // Determine the size of the static buffer we have
+                                static_buffer = buf.len();
+                                // Determine how much data we have already copied
+                                let copied_data = self.data_copied.get();
+
+                                data_len = d.as_ref().len();
+
+                                // If we have more data then the static buffer
+                                if data_len > copied_data {
+                                    // Copy the data into the static buffer
+                                    if data_len < (copied_data + static_buffer) {
```

Comment:
There could be an integer overflow in `copied_data + static_buffer`. How are we making sure that's not the case?

The following would be clearer IMO:

```rust
let remaining_data = &d.as_ref()[copied_data..];
let remaining_len = data_len - copied_data;

if remaining_len < static_buffer {
    buf[..remaining_len].copy_from_slice(remaining_data);
} else {
    buf.copy_from_slice(&remaining_data[..static_buffer]);
}
```